rfcs/0079-include-key-as.md

```diff
@@ -0,0 +1,89 @@
+  * **To Be Discussed**: whether we can afford to perform a breaking change for the above exception.
+* For all connectors with `format plain`, RisingWave derives the `_rw_key` column as a normal one
+  ([ref](https://github.com/risingwavelabs/risingwave/pull/13278)) at a POC request. This enforces an additional column
+  for all tables, which is not ideal.
+
+Aside from message keys, users may want to ingest the timestamp as well as the topic headers into the table for further
+analysis.
+
+This RFC proposes a new syntax to allow users to specify the column name for message components other than payload.
+
+## Proposed Syntax
+
+```sql
+create source/table (
+    ..,
+    primary key ( <key-column> )
+)
+    [ include key [ as <key-column> ] ]
+    [ include timestamp [ as <timestamp-column>] ]
+    [ include header [as <header-column>] ]
+    [ include ... [ as ... ] ]
+with ( ... )
+format ... encode ... ( ... )
+[ key encode [ ( ... ) ] ]
+```
+
+* If `as` name is not specified, a connector-component naming template will be applied
+  * For connector kafka and component key, the derived message key column name is `_rw_kafka_key`.
+* The default type for message key column is `bytea`. The priority of the type definition is: 
+  `key encode` > infer from `format ... encode ...` > default type  
+* **Important**: `include key` is required for `format upsert` and RisingWave will use the key column as one and 
+  only primary key to perform upsert semantic. It does not allow to specify multiple columns as primary key
+  even if they are part of the key.
+
+## Specifications
+
+### Kafka
+
+| Allowed Components | Default Type                               | Note                                              |
+|--------------------|--------------------------------------------|---------------------------------------------------|
+| key                | `bytea`                                    | Allow overwritten by `encode` and `key encode`    |
+| timestamp          | `timestamp with time zone` (i64 in millis) | Refer to `CreateTime` rather than `LogAppendTime` |
+| partition          | `i64`                                      | The message is from which partition               |
+| offset             | `i64`                                      | The offset in the partition                       |
+| header             | `struct<varchar, bytea>[]`                 | KV pairs along with message                       |
+
+### Pulsar
+
+| Allowed Components | Default Type | Note                                                                                      |
+|--------------------|--------------|-------------------------------------------------------------------------------------------|
+| key                | `bytea`      | Allow overwritten by `encode` and `key encode`. Refer to `MessageMetadata::partition_key` |
+
+More components are available at [here](https://docs.rs/pulsar/latest/pulsar/message/proto/struct.MessageMetadata.html).
```

Comment:
How do we decide which metadata fields to expose? Do we intend to expose similar (not sure if same) concepts from different connectors using the same word? For example
* kafka `key`, pulsar `partition_key`, kinesis `partition_key`
  * There is also pulsar `ordering_key` or kinesis [ExplicitHashKey](https://docs.aws.amazon.com/kinesis/latest/APIReference/API_PutRecord.html#API_PutRecord_RequestSyntax)
* kafka `offset`, pulsar `sequence_id`, kinesis `sequence_number`
  * i64 vs u64 vs string
* kafka `timestamp`, pulsar `publish_time`, kinesis `approximate_arrival_timestamp`
  * There is also pulsar `event_time`
* kafka `headers`, pulsar `properties`
  * kafka headers is `struct<varchar, bytea>[]` but pulsar properties is `struct<varchar, varchar>[]`

Reply:
> Do we intend to expose similar (not sure if same) concepts from different connectors using the same word?

I think it's not necessary to unify them. If there is ambiguity, I tend to let them have different names.

Reply:
Update: now we do want to unify `partition` and `offset` for them.. @tabVersion Can you add the rationale for that in the RFC?  😇

Reply:
Yes, we are going to unify the partition and offset column type for all connectors.
The basic reason is that we want to impl a source exec level throttling and requires a chunk to be cut anywhere with maintaining the offset info. Besides, the source backfill feature also rely on the behavior to tell when to end the backfill stage.